src/Pirouette/Transformations/Monomorphization.hs

```diff
@@ -27,15 +27,22 @@ import Pirouette.Monad
 import Pirouette.Term.Syntax
 import Pirouette.Term.Syntax.Subst
 import qualified Pirouette.Term.Syntax.SystemF as SystF
+import Pirouette.Transformations.Prenex (Prenex)
+import Pirouette.Transformations.Tagged
 import Pirouette.Utils
 
 -- * Monomorphization
 
+data Monomorphized
+
+monomorphize :: (Language lang) => Xform '[Prenex] '[Monomorphized] (PrtUnorderedDefs lang)
```

Comment:
I'm wondering here if this should be
`Xform '[Prenex] '[Prenex, Monomorphized]`
In other words, is it implicit that transformations cannot "undo" any previous transformation ?

Reply:
Yep, that's the assumption — transformations don't undo anything that happened before.

I agree this is not ideal, but I'm not sure there's a good, generic and fault-proof way to keep track of this: even if a transformation doesn't require something, it might accidentally break invariants established by some other transformation earlier.

If we care about such scenarios and if we think it's worth it, then the best approach would be to just do all the required transformations _inside_ the current one instead of keeping track of this "globally".

Reply:
For testing purposes, we could write dynamic checks for PNF / monomorphised code, etc. Basically a spec, which shouldn't be too hard to write. We could then check that transformations do indeed yield what they claim and check that other transformations do not violate the invariants. Of course, we'd only use that during testing and not when actually using the tool (although I suspect that such tests might be linear in the size of the inputs while symbolic exploration is exponential).